File: src/chess/movegen.ts
import { cloneBoard, opposite, squareName, type BoardState, type Color, type Piece } from './board.ts';
import type { Move } from './moveCodec.ts';

const KNIGHT = [[1, 2], [2, 1], [-1, 2], [-2, 1], [1, -2], [2, -1], [-1, -2], [-2, -1]];
const KING = [[1, 1], [1, 0], [1, -1], [0, 1], [0, -1], [-1, 1], [-1, 0], [-1, -1]];
const BISHOP = [[1, 1], [1, -1], [-1, 1], [-1, -1]];
const ROOK = [[1, 0], [-1, 0], [0, 1], [0, -1]];
// Hoisted: `[...BISHOP, ...ROOK]` inline rebuilt an 8-entry array of arrays for
// every queen, on every generation.
const QUEEN = [...BISHOP, ...ROOK];

const fileOf = (sq: number) => sq & 7;
const rankOf = (sq: number) => sq >> 3;
const on = (f: number, r: number) => f >= 0 && f < 8 && r >= 0 && r < 8;
const idx = (f: number, r: number) => f + (r << 3);
const colorOf = (piece: Piece | null): Color | null => piece?.[0] as Color | undefined ?? null;

/**
 * Piece literals per colour, hoisted out of the scan loops.
 *
 * `isSquareAttacked` used to build its comparison strings inline (`` `${by}n` ``),
 * which allocates once per square probed — and it probes on every legality test,
 * which runs once per pseudo-legal move. Same for `kingSquare`'s target. These
 * are the same eight strings forever, so build them once.
 */
const PIECES = {
  w: { p: 'wp', n: 'wn', b: 'wb', r: 'wr', q: 'wq', k: 'wk' },
  b: { p: 'bp', n: 'bn', b: 'bb', r: 'br', q: 'bq', k: 'bk' },
} as const;

function addStep(board: BoardState, moves: Move[], from: number, df: number, dr: number): void {
  const f = fileOf(from) + df, r = rankOf(from) + dr;
  if (!on(f, r)) return;
  const to = idx(f, r);
  if (colorOf(board.squares[to]) !== board.turn) moves.push({ from, to });
}

function addSlides(board: BoardState, moves: Move[], from: number, dirs: number[][]): void {
  for (const [df, dr] of dirs) {
    let f = fileOf(from) + df, r = rankOf(from) + dr;
    while (on(f, r)) {
      const to = idx(f, r);
      const occ = board.squares[to];
      if (!occ) moves.push({ from, to });
      else {
        if (colorOf(occ) !== board.turn) moves.push({ from, to });
        break;
      }
      f += df; r += dr;
    }
  }
}

function canCastle(board: BoardState, side: 'K' | 'Q' | 'k' | 'q'): boolean {
  if (!board.castling.includes(side)) return false;
  const white = side === 'K' || side === 'Q';
  const color: Color = white ? 'w' : 'b';
  const rank = white ? 0 : 7;
  const kingFrom = idx(4, rank);
  const rookFrom = side === 'K' || side === 'k' ? idx(7, rank) : idx(0, rank);
  const pieces = PIECES[color];
  if (board.turn !== color || board.squares[kingFrom] !== pieces.k || board.squares[rookFrom] !== pieces.r) return false;
  const between = side === 'K' || side === 'k' ? [idx(5, rank), idx(6, rank)] : [idx(1, rank), idx(2, rank), idx(3, rank)];
  if (between.some((sq) => board.squares[sq])) return false;
  const pass = side === 'K' || side === 'k' ? [idx(4, rank), idx(5, rank), idx(6, rank)] : [idx(4, rank), idx(3, rank), idx(2, rank)];
  return pass.every((sq) => !isSquareAttacked(board, sq, opposite(color)));
}

export function pseudoLegalMoves(board: BoardState): Move[] {
  const moves: Move[] = [];
  for (let from = 0; from < 64; from++) {
    const piece = board.squares[from];
    if (!piece || colorOf(piece) !== board.turn) continue;
    const role = piece[1];
    if (role === 'p') {
      const dir = board.turn === 'w' ? 1 : -1;
      const startRank = board.turn === 'w' ? 1 : 6;
      const promoteRank = board.turn === 'w' ? 7 : 0;
      const f = fileOf(from), r = rankOf(from);
      const oneR = r + dir;
      if (on(f, oneR) && !board.squares[idx(f, oneR)]) {
        const to = idx(f, oneR);
        if (oneR === promoteRank) for (const promotion of ['q', 'r', 'b', 'n'] as const) moves.push({ from, to, promotion });
        else moves.push({ from, to });
        const twoR = r + 2 * dir;
        if (r === startRank && on(f, twoR) && !board.squares[idx(f, twoR)]) moves.push({ from, to: idx(f, twoR) });
      }
      for (const df of [-1, 1]) {
        const cf = f + df;
        if (!on(cf, oneR)) continue;
        const to = idx(cf, oneR);
        if (colorOf(board.squares[to]) === opposite(board.turn) || to === board.epSquare) {
          if (oneR === promoteRank) for (const promotion of ['q', 'r', 'b', 'n'] as const) moves.push({ from, to, promotion });
          else moves.push({ from, to });
        }
      }
    } else if (role === 'n') { for (const [df, dr] of KNIGHT) addStep(board, moves, from, df, dr); }
    else if (role === 'b') addSlides(board, moves, from, BISHOP);
    else if (role === 'r') addSlides(board, moves, from, ROOK);
    else if (role === 'q') addSlides(board, moves, from, QUEEN);
    else if (role === 'k') {
      for (const [df, dr] of KING) addStep(board, moves, from, df, dr);
      if (canCastle(board, board.turn === 'w' ? 'K' : 'k')) moves.push({ from, to: idx(6, board.turn === 'w' ? 0 : 7) });
      if (canCastle(board, board.turn === 'w' ? 'Q' : 'q')) moves.push({ from, to: idx(2, board.turn === 'w' ? 0 : 7) });
    }
  }
  return moves;
}

export function isSquareAttacked(board: BoardState, square: number, by: Color): boolean {
  const f = fileOf(square), r = rankOf(square);
  const pieces = PIECES[by];
  const squares = board.squares;
  const pawnFromRank = r + (by === 'w' ? -1 : 1);
  for (const df of [-1, 1]) {
    const pf = f + df;
    if (on(pf, pawnFromRank) && squares[idx(pf, pawnFromRank)] === pieces.p) return true;
  }
  for (const [df, dr] of KNIGHT) {
    const nf = f + df, nr = r + dr;
    if (on(nf, nr) && squares[idx(nf, nr)] === pieces.n) return true;
  }
  for (const [df, dr] of BISHOP) {
    let sf = f + df, sr = r + dr;
    while (on(sf, sr)) {
      const piece = squares[idx(sf, sr)];
      if (piece) { if (piece === pieces.b || piece === pieces.q) return true; break; }
      sf += df; sr += dr;
    }
  }
  for (const [df, dr] of ROOK) {
    let sf = f + df, sr = r + dr;
    while (on(sf, sr)) {
      const piece = squares[idx(sf, sr)];
      if (piece) { if (piece === pieces.r || piece === pieces.q) return true; break; }
      sf += df; sr += dr;
    }
  }
  for (const [df, dr] of KING) {
    const kf = f + df, kr = r + dr;
    if (on(kf, kr) && squares[idx(kf, kr)] === pieces.k) return true;
  }
  return false;
}

export function kingSquare(board: BoardState, color: Color): number | null {
  const target = PIECES[color].k;
  const squares = board.squares;
  for (let sq = 0; sq < 64; sq += 1) if (squares[sq] === target) return sq;
  return null;
}

export function inCheck(board: BoardState, color: Color = board.turn): boolean {
  const king = kingSquare(board, color);
  return king === null ? true : isSquareAttacked(board, king, opposite(color));
}

function moveKeepsKingSafe(board: BoardState, move: Move, color: Color): boolean {
  const squares = board.squares;
  const piece = squares[move.from];
  if (!piece) return false;

  const captured = squares[move.to];
  const isPawn = piece[1] === 'p';
  const isEnPassant = isPawn
    && move.to === board.epSquare
    && !captured
    && fileOf(move.from) !== fileOf(move.to);
  const enPassantCaptureSquare = isEnPassant ? idx(fileOf(move.to), rankOf(move.from)) : -1;
  const enPassantCaptured = enPassantCaptureSquare >= 0 ? squares[enPassantCaptureSquare] : null;

  const isCastle = piece[1] === 'k' && Math.abs(fileOf(move.to) - fileOf(move.from)) === 2;
  const castleRank = color === 'w' ? 0 : 7;
  const rookFrom = isCastle ? idx(fileOf(move.to) === 6 ? 7 : 0, castleRank) : -1;
  const rookTo = isCastle ? idx(fileOf(move.to) === 6 ? 5 : 3, castleRank) : -1;
  const rookFromPiece = rookFrom >= 0 ? squares[rookFrom] : null;
  const rookToPiece = rookTo >= 0 ? squares[rookTo] : null;

  squares[move.from] = null;
  if (enPassantCaptureSquare >= 0) squares[enPassantCaptureSquare] = null;
  squares[move.to] = move.promotion ? PIECES[color][move.promotion] : piece;
  if (rookFrom >= 0) {
    squares[rookTo] = rookFromPiece;
    squares[rookFrom] = null;
  }

  try {
    return !inCheck(board, color);
  } finally {
    if (rookFrom >= 0) {
      squares[rookFrom] = rookFromPiece;
      squares[rookTo] = rookToPiece;
    }
    squares[move.from] = piece;
    squares[move.to] = captured;
    if (enPassantCaptureSquare >= 0) squares[enPassantCaptureSquare] = enPassantCaptured;
  }
}

export function legalMoves(board: BoardState): Move[] {
  const moves = pseudoLegalMoves(board);
  const color = board.turn;
  let legalCount = 0;
  for (const move of moves) {
    if (moveKeepsKingSafe(board, move, color)) moves[legalCount++] = move;
  }
  moves.length = legalCount;
  return moves;
}

function castlingWithout(castling: string, removed: string): string {
  const next = castling.split('').filter((ch) => !removed.includes(ch)).join('');
  return next || '-';
}

export function makeMove(board: BoardState, move: Move): BoardState {
  const next = cloneBoard(board);
  const piece = next.squares[move.from];
  if (!piece) throw new Error('No piece on source square');
  const captured = next.squares[move.to];
  const fromName = squareName(move.from), toName = squareName(move.to);
  const isPawn = piece[1] === 'p';
  const isCastle = piece[1] === 'k' && Math.abs(fileOf(move.to) - fileOf(move.from)) === 2;
  const isEp = isPawn && move.to === board.epSquare && !captured && fileOf(move.from) !== fileOf(move.to);

  next.squares[move.from] = null;
  if (isEp) next.squares[idx(fileOf(move.to), rankOf(move.from))] = null;
  next.squares[move.to] = move.promotion ? `${board.turn}${move.promotion}` as Piece : piece;
  if (isCastle) {
    const rank = board.turn === 'w' ? 0 : 7;
    if (fileOf(move.to) === 6) {
      next.squares[idx(5, rank)] = next.squares[idx(7, rank)];
      next.squares[idx(7, rank)] = null;
    } else {
      next.squares[idx(3, rank)] = next.squares[idx(0, rank)];
      next.squares[idx(0, rank)] = null;
    }
  }

  let castling = board.castling === '-' ? '' : board.castling;
  if (piece === 'wk') castling = castlingWithout(castling, 'KQ');
  if (piece === 'bk') castling = castlingWithout(castling, 'kq');
  if (fromName === 'a1' || toName === 'a1') castling = castlingWithout(castling, 'Q');
  if (fromName === 'h1' || toName === 'h1') castling = castlingWithout(castling, 'K');
  if (fromName === 'a8' || toName === 'a8') castling = castlingWithout(castling, 'q');
  if (fromName === 'h8' || toName === 'h8') castling = castlingWithout(castling, 'k');
  next.castling = castling || '-';

  next.turn = opposite(board.turn);
  next.epSquare = isPawn && Math.abs(rankOf(move.to) - rankOf(move.from)) === 2 ? idx(fileOf(move.from), (rankOf(move.from) + rankOf(move.to)) / 2) : null;
  next.halfmove = isPawn || captured || isEp ? 0 : board.halfmove + 1;
  next.fullmove += board.turn === 'b' ? 1 : 0;
  return next;
}

File: rust/tiny_leela_core/src/movegen.rs
use crate::board::{
    file, idx, on, rank, Board, Color, Move, Piece, Role, BISHOP, KING, KNIGHT, ROOK,
};

fn push_step(board: &Board, moves: &mut Vec<Move>, from: u8, df: i8, dr: i8) {
    let (f, r) = (file(from) + df, rank(from) + dr);
    if !on(f, r) {
        return;
    }
    let to = idx(f, r);
    if board.squares[to as usize].map(|p| p.color) != Some(board.turn) {
        moves.push(Move {
            from,
            to,
            promotion: None,
        });
    }
}

fn push_slides(board: &Board, moves: &mut Vec<Move>, from: u8, dirs: &[(i8, i8)]) {
    for &(df, dr) in dirs {
        let (mut f, mut r) = (file(from) + df, rank(from) + dr);
        while on(f, r) {
            let to = idx(f, r);
            match board.squares[to as usize] {
                None => moves.push(Move {
                    from,
                    to,
                    promotion: None,
                }),
                Some(piece) => {
                    if piece.color != board.turn {
                        moves.push(Move {
                            from,
                            to,
                            promotion: None,
                        });
                    }
                    break;
                }
            }
            f += df;
            r += dr;
        }
    }
}

fn can_castle(board: &Board, mask: u8) -> bool {
    if board.castling & mask == 0 {
        return false;
    }
    let white = mask == 1 || mask == 2;
    let color = if white { Color::White } else { Color::Black };
    let r = if white { 0 } else { 7 };
    let king_from = idx(4, r);
    let rook_from = if mask == 1 || mask == 4 {
        idx(7, r)
    } else {
        idx(0, r)
    };
    if board.turn != color
        || board.squares[king_from as usize]
            != Some(Piece {
                color,
                role: Role::King,
            })
        || board.squares[rook_from as usize]
            != Some(Piece {
                color,
                role: Role::Rook,
            })
    {
        return false;
    }
    let between: &[u8] = if mask == 1 || mask == 4 {
        &[idx(5, r), idx(6, r)]
    } else {
        &[idx(1, r), idx(2, r), idx(3, r)]
    };
    if between.iter().any(|&s| board.squares[s as usize].is_some()) {
        return false;
    }
    let pass: &[u8] = if mask == 1 || mask == 4 {
        &[idx(4, r), idx(5, r), idx(6, r)]
    } else {
        &[idx(4, r), idx(3, r), idx(2, r)]
    };
    pass.iter()
        .all(|&s| !is_square_attacked(board, s, color.opposite()))
}

pub fn pseudo_legal_moves(board: &Board) -> Vec<Move> {
    let mut moves = Vec::with_capacity(64);
    for from in 0u8..64 {
        let Some(piece) = board.squares[from as usize] else {
            continue;
        };
        if piece.color != board.turn {
            continue;
        }
        match piece.role {
            Role::Pawn => {
                let dir = if board.turn == Color::White { 1 } else { -1 };
                let start = if board.turn == Color::White { 1 } else { 6 };
                let promo = if board.turn == Color::White { 7 } else { 0 };
                let (f, r) = (file(from), rank(from));
                let one = r + dir;
                if on(f, one) && board.squares[idx(f, one) as usize].is_none() {
                    let to = idx(f, one);
                    if one == promo {
                        for promotion in [Role::Queen, Role::Rook, Role::Bishop, Role::Knight] {
                            moves.push(Move {
                                from,
                                to,
                                promotion: Some(promotion),
                            });
                        }
                    } else {
                        moves.push(Move {
                            from,
                            to,
                            promotion: None,
                        });
                    }
                    let two = r + 2 * dir;
                    if r == start && on(f, two) && board.squares[idx(f, two) as usize].is_none() {
                        moves.push(Move {
                            from,
                            to: idx(f, two),
                            promotion: None,
                        });
                    }
                }
                for df in [-1, 1] {
                    let cf = f + df;
                    if !on(cf, one) {
                        continue;
                    }
                    let to = idx(cf, one);
                    let capture = board.squares[to as usize].map(|p| p.color)
                        == Some(board.turn.opposite())
                        || Some(to) == board.ep_square;
                    if capture {
                        if one == promo {
                            for promotion in [Role::Queen, Role::Rook, Role::Bishop, Role::Knight] {
                                moves.push(Move {
                                    from,
                                    to,
                                    promotion: Some(promotion),
                                });
                            }
                        } else {
                            moves.push(Move {
                                from,
                                to,
                                promotion: None,
                            });
                        }
                    }
                }
            }
            Role::Knight => {
                for &(df, dr) in KNIGHT {
                    push_step(board, &mut moves, from, df, dr);
                }
            }
            Role::Bishop => push_slides(board, &mut moves, from, BISHOP),
            Role::Rook => push_slides(board, &mut moves, from, ROOK),
            Role::Queen => {
                push_slides(board, &mut moves, from, BISHOP);
                push_slides(board, &mut moves, from, ROOK);
            }
            Role::King => {
                for &(df, dr) in KING {
                    push_step(board, &mut moves, from, df, dr);
                }
                if can_castle(board, if board.turn == Color::White { 1 } else { 4 }) {
                    moves.push(Move {
                        from,
                        to: idx(6, if board.turn == Color::White { 0 } else { 7 }),
                        promotion: None,
                    });
                }
                if can_castle(board, if board.turn == Color::White { 2 } else { 8 }) {
                    moves.push(Move {
                        from,
                        to: idx(2, if board.turn == Color::White { 0 } else { 7 }),
                        promotion: None,
                    });
                }
            }
        }
    }
    moves
}

pub fn is_square_attacked(board: &Board, sq: u8, by: Color) -> bool {
    let (f, r) = (file(sq), rank(sq));
    let pawn_rank = r + if by == Color::White { -1 } else { 1 };
    for df in [-1, 1] {
        let pf = f + df;
        if on(pf, pawn_rank)
            && board.squares[idx(pf, pawn_rank) as usize]
                == Some(Piece {
                    color: by,
                    role: Role::Pawn,
                })
        {
            return true;
        }
    }
    for &(df, dr) in KNIGHT {
        let (nf, nr) = (f + df, r + dr);
        if on(nf, nr)
            && board.squares[idx(nf, nr) as usize]
                == Some(Piece {
                    color: by,
                    role: Role::Knight,
                })
        {
            return true;
        }
    }
    for &(df, dr) in BISHOP {
        if ray_attacked(board, f, r, df, dr, by, Role::Bishop) {
            return true;
        }
    }
    for &(df, dr) in ROOK {
        if ray_attacked(board, f, r, df, dr, by, Role::Rook) {
            return true;
        }
    }
    for &(df, dr) in KING {
        let (kf, kr) = (f + df, r + dr);
        if on(kf, kr)
            && board.squares[idx(kf, kr) as usize]
                == Some(Piece {
                    color: by,
                    role: Role::King,
                })
        {
            return true;
        }
    }
    false
}

fn ray_attacked(board: &Board, f: i8, r: i8, df: i8, dr: i8, by: Color, slider: Role) -> bool {
    let (mut sf, mut sr) = (f + df, r + dr);
    while on(sf, sr) {
        if let Some(p) = board.squares[idx(sf, sr) as usize] {
            return p.color == by && (p.role == slider || p.role == Role::Queen);
        }
        sf += df;
        sr += dr;
    }
    false
}

pub fn king_square(board: &Board, color: Color) -> Option<u8> {
    board
        .squares
        .iter()
        .position(|&p| {
            p == Some(Piece {
                color,
                role: Role::King,
            })
        })
        .map(|i| i as u8)
}

pub fn in_check(board: &Board, color: Color) -> bool {
    king_square(board, color)
        .map(|sq| is_square_attacked(board, sq, color.opposite()))
        .unwrap_or(true)
}

pub fn legal_moves(board: &Board) -> Vec<Move> {
    pseudo_legal_moves(board)
        .into_iter()
        .filter(|&m| !in_check(&make_move(board, m), board.turn))
        .collect()
}

fn remove_castling(c: &mut u8, mask: u8) {
    *c &= !mask;
}

pub fn make_move(board: &Board, m: Move) -> Board {
    let mut next = board.clone();
    let piece = next.squares[m.from as usize].expect("No piece on source square");
    let captured = next.squares[m.to as usize];
    let is_pawn = piece.role == Role::Pawn;
    let is_castle = piece.role == Role::King && (file(m.to) - file(m.from)).abs() == 2;
    let is_ep = is_pawn
        && Some(m.to) == board.ep_square
        && captured.is_none()
        && file(m.from) != file(m.to);
    next.squares[m.from as usize] = None;
    if is_ep {
        next.squares[idx(file(m.to), rank(m.from)) as usize] = None;
    }
    next.squares[m.to as usize] = Some(Piece {
        color: board.turn,
        role: m.promotion.unwrap_or(piece.role),
    });
    if is_castle {
        let r = if board.turn == Color::White { 0 } else { 7 };
        if file(m.to) == 6 {
            next.squares[idx(5, r) as usize] = next.squares[idx(7, r) as usize];
            next.squares[idx(7, r) as usize] = None;
        } else {
            next.squares[idx(3, r) as usize] = next.squares[idx(0, r) as usize];
            next.squares[idx(0, r) as usize] = None;
        }
    }
    match piece {
        Piece {
            color: Color::White,
            role: Role::King,
        } => remove_castling(&mut next.castling, 1 | 2),
        Piece {
            color: Color::Black,
            role: Role::King,
        } => remove_castling(&mut next.castling, 4 | 8),
        _ => {}
    }
    for (sq, mask) in [(0, 2), (7, 1), (56, 8), (63, 4)] {
        if m.from == sq || m.to == sq {
            remove_castling(&mut next.castling, mask);
        }
    }
    next.turn = board.turn.opposite();
    next.ep_square = if is_pawn && (rank(m.to) - rank(m.from)).abs() == 2 {
        Some(idx(file(m.from), (rank(m.from) + rank(m.to)) / 2))
    } else {
        None
    };
    next.halfmove = if is_pawn || captured.is_some() || is_ep {
        0
    } else {
        board.halfmove + 1
    };
    next.fullmove = board.fullmove + if board.turn == Color::Black { 1 } else { 0 };
    next
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{move_to_uci, parse_fen, Piece, START_FEN};

    #[test]
    fn start_position_has_20_legal_moves() {
        assert_eq!(legal_moves(&parse_fen(START_FEN).unwrap()).len(), 20);
    }

    #[test]
    fn rejects_king_into_check() {
        let b = parse_fen("k3r3/8/8/8/8/8/8/4K3 w - - 0 1").unwrap();
        let moves: Vec<String> = legal_moves(&b).into_iter().map(move_to_uci).collect();
        assert!(!moves.contains(&"e1e2".to_string()));
    }

    #[test]
    fn castling_and_ep() {
        let b = parse_fen("r3k2r/8/8/8/8/8/8/R3K2R w KQkq - 0 1").unwrap();
        let moves: Vec<String> = legal_moves(&b).into_iter().map(move_to_uci).collect();
        assert!(moves.contains(&"e1g1".to_string()));
        assert!(moves.contains(&"e1c1".to_string()));
        let ep = make_move(
            &parse_fen("8/8/8/3pP3/8/8/8/4K2k w - d6 0 1").unwrap(),
            Move {
                from: 36,
                to: 43,
                promotion: None,
            },
        );
        assert!(ep.squares[35].is_none());
        assert_eq!(
            ep.squares[43],
            Some(Piece {
                color: Color::White,
                role: Role::Pawn
            })
        );
    }
}
